src/ORM/DataList.php

```diff
@@ -1308,13 +1317,19 @@ private function fetchEagerLoadManyMany(
         $relationListClass = get_class($relationList);
 
         // Get ALL of the items for this relation up front, for ALL of the parents
-        $fetchedRows = $relationList->forForeignID($parentIDs)->getFinalisedQuery();
+        $fetchList = $relationList->forForeignID($parentIDs);
+        $fetchList = $this->applyPreFilteringToEagerloading($fetchList, $relationChain, $relationType);
+        $fetchedRows = $fetchList->getFinalisedQuery();
 
         foreach ($fetchedRows as $row) {
             $fetchedRowsArray[$row['ID']] = $row;
             $fetchedIDs[] = $row['ID'];
         }
 
+        // Get the join records so we can correctly identify which children belong to which parents
+        // This also holds extra fields data
+        $joinRows = DB::query('SELECT * FROM "' . $joinTable . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ') AND ' . $childIDField . ' IN (' . implode(',', $fetchedIDs) . ')');
```

Comment:
Note we're now only getting the join records that are applicable to the potentially-filtered-down set of relations.